hal/explicit_pinmap.cpp

```diff
@@ -23,6 +23,8 @@
 #include "analogout_api.h"
 #include "i2c_api.h"
 #include "serial_api.h"
+#define CONSTEXPR constexpr
```

Comment:
As this is local trickery, I'd be inclined to call it `PINMAP_CONSTEXPR` or something, to make sure not to collide with someone else doing something similar.

Alternatively, maybe you could incorporate something like this into `<mstd_cstddef>`? There are some "C++11 or C++14" macros there - maybe we could have "C++11 or C99/C++98" macros to go with them? Make that header work in C mode, and define `MSTD_CONSTEXPR_OBJ` as `const` or `constexpr` accordingly.

Reply:
Although, I don't think you technically **need** to actually mark those tables as `constexpr` anyway. As long as they are constant expressions, these function definitions should work and be `constexpr`.

Putting the `constexpr` in the table just forces a pre-check - you'd get a compile error from the tables themselves if you put something non-constant in.

I guess it's kind of nice to force the check - with just `const` there is a chance the table silently goes into RAM if not a constant expression, but `constexpr` blocks it. But we've managed without that check up until now, so I wouldn't sweat it too much.

Reply:
You're right. GCC does accept it, but clang refuses, and it's allowed to.

Only `const int`s or enumerations can be used in constant expressions - you can't actually access an array of ints. That's one of those very early C++ things where they treated `const int` specially to try to get people to stop using #defines.

Reference: https://stackoverflow.com/questions/45593044/the-value-of-a-const-variable-is-or-is-not-usable-in-a-constant-expression-depe

Reply:
I was misled by the answer here: https://stackoverflow.com/questions/14116003/difference-between-constexpr-and-const

because I missed the clause "of integral or enumeration type" in the bit about "an object can be used as constant expression without being declared constexpr".

Reply:
Fixed. Used MSTD_CONSTEXPR_OBJ_C14 macro.